### PR TITLE
[Feature] Authentication and DB configuration

### PR DESCRIPTION
**Pull request by umar0932**
- Graphql Configured
- Typeorm and Postgres configured
- Husky configured
- Separate Environments development and production configeration
- Linter and prettier configured
- Absolute paths are configured
- User Model Created (will be refined further)
- User Module Created
- Authentication of User (will be improved further)
- Authorization  of User on basis of roles implemented (will be improved further)
- Becrytion password
- Dockerized postgres (will be improved further)

### CODE REVIEW

File: src/modules/enums/roles.enum.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+enum Role {
+  USER = 'USER',
+  ADMIN = 'ADMIN'
+}
+
+export default Role
```
Comment (Usama4722) [inline src/modules/enums/roles.enum.ts:6]: Roles are attendee and organizer. Admin role is a separate entity,

File: src/modules/users/entities/user.entity.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  password: string
+
+  @Column({
+    type: 'enum',
+    enum: Role,
+    default: Role.USER
+  })
+  @Field()
+  role: Role
+}
```
Comment (Usama4722) [inline src/modules/users/entities/user.entity.ts:33]: add additional fields must required. like firstname, lastname, authProviders, account status, last login, created date, last updated date, stripe user id.

File: src/schema.gql [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# ------------------------------------------------------
+# THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
+# ------------------------------------------------------
+
+input CreateUserInput {
+  email: String!
+  password: String!
+  username: String!
+}
```
Comment (Usama4722) [inline src/schema.gql:9]: username is not in user model.

File: src/schema.gql [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  users: [User!]!
+}
+
+type User {
+  email: String!
+  id: Int!
+  password: String!
+  role: String!
+  username: String!
+}
```
Comment (Usama4722) [inline src/schema.gql:39]: username is not in user model.

File: test/setEnvVars.js [hunk 5]
```diff
@@ -0,0 +1,2 @@
+const dotenv = require('dotenv');
+dotenv.config({ path: './env.test' });
```
Comment (Usama4722) [inline test/setEnvVars.js:2]: what is this?

File: src/modules/users/users.resolver.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import { Resolver, Query, Mutation, Args } from '@nestjs/graphql';
+import { UseGuards } from '@nestjs/common';
+
+import { CreateUserInput } from './dto/create-user.input';
+import { JwtAuthGuard } from '@app/auth/guards/jwt-auth.guard';
+import Role from '@app/enums/roles.enum';
+import { Roles } from '@app/auth/decorators/roles.decorator';
+import { RolesGuard } from '@app/auth/guards/roles.guard';
+import { User } from './entities/user.entity';
+import { UsersService } from './users.service';
```
Comment (Usama4722) [inline src/modules/users/users.resolver.ts:10]: It should be user service not users service.

File: src/modules/auth/auth.service.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import { Injectable } from '@nestjs/common'
+import { JwtService } from '@nestjs/jwt'
+
+import * as bcrypt from 'bcrypt'
+
+import { CreateUserInput } from '@app/users/dto/create-user.input'
+import { LoginUserInput } from './dto/login-user.input'
+import { UsersService } from '@app/users/users.service'
+
```
Comment (Usama4722) [inline src/modules/auth/auth.service.ts:9]: should user service not users service.

File: src/modules/users/users.resolver.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  // Make sure to add RolesGuard to the @UseGuards() decorator
+  @UseGuards(JwtAuthGuard, RolesGuard)
+  // Create roles in enums/roles.enum.ts
+  // Import the enum
+  // Add the right roles to the @Roles() decorator
+  @Roles(Role.ADMIN)
+  findAll(): Promise<User[]> {
+    return this.usersService.findAll();
+  }
+
```
Comment (Usama4722) [inline src/modules/users/users.resolver.ts:27]: Role should be organizer.

**Review by Usama4722 — CHANGES_REQUESTED** [review]

File: src/app.module.ts [hunk 9]
```diff
@@ -1,10 +1,58 @@
...
+import { join } from 'path'
+
+import { AuthModule } from './modules/auth/auth.module'
 import { AppService } from './app.service'
+import { AppResolver } from './app.resolver'
+import { UsersModule } from './modules/users/users.module'
+
+dotenv.config()
+
+const env = `${(process.env.NODE_ENV || 'development').toLowerCase()}`
```
[the remark this answers is not in the document]
Comment (umar0932) [reply src/app.module.ts:17]: Issue resolved

File: src/modules/enums/roles.enum.ts (hunk 1 above)
Comment (umar0932) [reply src/modules/enums/roles.enum.ts:6]: Issue Resolved

File: src/schema.gql (hunk 3 above)
Comment (umar0932) [reply src/schema.gql:9]: Comment fixed

File: src/schema.gql (hunk 4 above)
Comment (umar0932) [reply src/schema.gql:39]: fixed

File: test/setEnvVars.js (hunk 5 above)
Comment (umar0932) [reply test/setEnvVars.js:2]: fixed

File: src/modules/users/users.resolver.ts (hunk 6 above)
Comment (umar0932) [reply src/modules/users/users.resolver.ts:10]: fixed

File: src/modules/auth/auth.service.ts (hunk 7 above)
Comment (umar0932) [reply src/modules/auth/auth.service.ts:9]: fixed

File: src/modules/users/users.resolver.ts (hunk 8 above)
Comment (umar0932) [reply src/modules/users/users.resolver.ts:27]: It should be the Admin as admin can see all users

File: src/modules/enums/roles.enum.ts [hunk 10]
```diff
@@ -1,5 +1,6 @@
 enum Role {
   USER = 'USER',
+  ORGANIZER = 'ORGANIZER',
   ADMIN = 'ADMIN'
 }
```
Comment (Usama4722) [inline src/modules/enums/roles.enum.ts:6]: Admin roles will treat differently. user roles will be 2 attendee and organizer.

**Review by Usama4722 — CHANGES_REQUESTED** [review]

**Review by Usama4722 — DISMISSED** [review]

**Review by Usama4722 — APPROVED** [review]

**Pull request merged**